### DISCUSSION TITLE
How to skip hydration in SPA mode?

### DISCUSSION THREAD

**Discussion by dtinth** [remix-run/remix#8960] [Q&A]
I am using [MUI](https://mui.com/) in a Remix SPA app I’m developing.

I found that when our user has some Chrome extensions installed, hydration would fail, leading to errors like this and styles not working.

<img width="1394" alt="image" src="https://github.com/remix-run/remix/assets/193136/badb8957-ce99-45a0-b92c-a7e8921d1195">

When hydration is working properly, it would look like this:

<img width="1394" alt="image" src="https://github.com/remix-run/remix/assets/193136/a615bd40-f4c4-422a-8d48-a4f0fd806a6e">

[This “gotcha” is documented in Remix’s docs:](https://remix.run/docs/en/main/guides/gotchas#browser-extensions-injecting-code)

> <img width="814" alt="image" src="https://github.com/remix-run/remix/assets/193136/e944f9af-c1bb-465f-94c2-a969daa0a0ed">

However, it isn’t practical for us to ask our users to always use our app in incognito mode. Each user may have different extensions installed, and that’s out of our control.

In order to avoid [SSR hydration-related headache](https://twitter.com/danielberg_/status/1763341446143234462/photo/1), is it possible to tell Remix to _not_ hydrate the HTML, and instead, do something like what Create React App does, i.e., mounting into `#root` without hydration?

**Comment by robertjpayne** [comment]
@dtinth Pretty sure you can take your entry.client.tsx and do a renderRoot instead of hydrateRoot.

**Reply by dtinth** [reply]
Here’s how I did it:

```ts
// root.tsx
import {
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
} from "@remix-run/react";
import "./index.css";

export function Layout({ children }: { children: React.ReactNode }) {
  if (typeof window === "undefined") {
    return (
      <html lang="en" data-bs-theme="dark" className="dtinth">
        <head>
          <meta charSet="utf-8" />
          <meta name="viewport" content="width=device-width, initial-scale=1" />
          <Meta />
          <Links />
        </head>
        <body>
          <div id="app">{children}</div>
          <ScrollRestoration />
          <Scripts />
        </body>
      </html>
    );
  } else {
    return <>{children}</>;
  }
}
```

```
// entry.client.tsx
import { RemixBrowser } from "@remix-run/react";
import { startTransition, StrictMode } from "react";
import { createRoot } from "react-dom/client";

startTransition(() => {
  createRoot(document.getElementById("app")!).render(
    <StrictMode>
      <RemixBrowser />
    </StrictMode>
  );
});
```

So far it works, but not sure if there are better ways.

**Reply by Narixius** [reply]
It seems not the correct way to do that. In this case, the `<Meta />` component will not render on the page during development, and there is no control over meta tags using the `meta` function.
